internal/fxlog/fxlog.go

```diff
@@ -21,86 +21,127 @@
+type StartFailureError struct{ Err error }
+
+// StopSignalEvent is emitted whenever application receives a signal after
+// starting the application.
+type StopSignalEvent struct{ Signal string }
+
+// StopErrorEvent is emitted whenever we fail to stop cleanly.
+type StopErrorEvent struct{ Err error }
+
+// StartErrorEvent is emitted whenever a service fails to start.
+type StartErrorEvent struct{ Err error }
+
+// StartRollbackError is emitted whenever we fail to rollback cleanly after
+// a start error.
+type StartRollbackError struct{ Err error }
+
+// RunningEvent is emitted whenever an application is started successfully.
+type RunningEvent struct{}
+
 var _ Logger = (*zapLogger)(nil)
 
 type zapLogger struct {
 	logger *zap.Logger
 }
 
-func encodeFields(fields []Field, stack string) []zap.Field {
-	var fs []zap.Field
-	for _, field := range fields {
-		fs = append(fs, zap.Any(field.Key, field.Value))
-	}
-	if stack != "" {
-		fs = append(fs, zap.String("stack", stack))
-	}
-
-	return fs
-}
-
-func (l *zapLogger) Log(entry Entry) {
-	switch entry.Level {
-	case InfoLevel:
-		l.logger.Info(entry.Message, encodeFields(entry.Fields, entry.Stack)...)
-	case ErrorLevel:
-		l.logger.Error(entry.Message, encodeFields(entry.Fields, entry.Stack)...)
+func (l *zapLogger) LogEvent(event Event) {
+	switch e := event.(type) {
+	case LifecycleOnStartEvent:
+		l.logger.Info("starting", zap.String("caller", e.Caller))
+	case LifecycleOnStopEvent:
+		l.logger.Info("stopping", zap.String("caller", e.Caller))
+	case ApplyOptionsError:
+		l.logger.Error("error encountered while applying options", zap.Error(e.Err))
+	case SupplyEvent:
+		for _, rtype := range fxreflect.ReturnTypes(e.Constructor) {
+			l.logger.Info("supplying",
+				zap.String("constructor", fxreflect.FuncName(e.Constructor)),
+				zap.String("type", rtype),
+			)
+		}
+	case ProvideEvent:
+		for _, rtype := range fxreflect.ReturnTypes(e.Constructor) {
+			l.logger.Info("providing",
+				zap.String("constructor", fxreflect.FuncName(e.Constructor)),
+				zap.String("type", rtype),
+			)
+		}
+	case InvokeEvent:
+		l.logger.Info("invoke", zap.String("function", fxreflect.FuncName(e.Function)))
+	case InvokeFailedEvent:
+		l.logger.Error("fx.Invoke failed",
+			zap.Error(e.Err),
+			zap.String("stack", e.Stack.String()),
+			zap.String("function", fxreflect.FuncName(e.Function)))
+	case StartFailureError:
+		l.logger.Error("failed to start", zap.Error(event.(StartFailureError).Err))
+	case StopSignalEvent:
+		l.logger.Info("received signal", zap.String("signal", strings.ToUpper(e.Signal)))
+	case StopErrorEvent:
+		l.logger.Error("failed to stop cleanly", zap.Error(event.(StopErrorEvent).Err))
+	case StartRollbackError:
+		l.logger.Error("could not rollback cleanly", zap.Error(event.(StartRollbackError).Err))
+	case StartErrorEvent:
+		l.logger.Error("startup failed, rolling back", zap.Error(event.(StartErrorEvent).Err))
```

Comment:
Since we're doing `e := event.(type)`, we can use `e` here to reference the variable with the matched type instead of casting `event` again.
```suggestion
		l.logger.Error("failed to start", zap.Error(e.Err))
	case StopSignalEvent:
		l.logger.Info("received signal", zap.String("signal", strings.ToUpper(e.Signal)))
	case StopErrorEvent:
		l.logger.Error("failed to stop cleanly", zap.Error(e.Err))
	case StartRollbackError:
		l.logger.Error("could not rollback cleanly", zap.Error(e.Err))
	case StartErrorEvent:
		l.logger.Error("startup failed, rolling back", zap.Error(e.Err))
```